### PR TITLE
Bugfix/build worker engine plugin

### PR DESCRIPTION
**Pull request by joshuahuburn**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
Updated BuildWorker.bat to work with Engine plugin setup by default. Did some cleanup in the script while I was here too.
This change does not affect the project plugin build in any way.
We simply use the arguments already passed to BuildWorker.bat to locate the correct project directory, spatial directory and the build.exe.

#### Tests
Building with only an engine plugin works.

How can this be verified by QA?: Build the project using an engine plugin for the GDK.

#### Primary reviewers
@m-samiec @improbable-valentyn

### CODE REVIEW

File: SpatialGDK/Build/Scripts/BuildWorker.bat [hunk 1]
```diff
@@ -1,22 +1,63 @@
 @echo off
 
-rem Generate worker configs
-pushd "%~dp0..\..\..\..\..\..\spatial"
+rem Input args: BuildName, Platform, BuildConfiguration, uproject 
+
+rem If we are running as a project plugin then this will be the path to the spatial directory.
+:BuildAsProjectPlugin
+if exist "%~dp0..\..\..\..\..\..\spatial" (
```
Comment (improbable-valentyn) [inline SpatialGDK/Build/Scripts/BuildWorker.bat]: If the engine root is the directory just above the project root, won't this relative path be the same for both project plugin and the engine plugin?

Comment (joshuahuburn) [reply SpatialGDK/Build/Scripts/BuildWorker.bat]: This is unfortunately not true in all cases. You can use an Engine plugin but with the project located in any directory.

Comment (improbable-valentyn) [reply SpatialGDK/Build/Scripts/BuildWorker.bat]: What I mean is that the relative path `..\..\..\..\..\..\spatial` resolving to a valid directory does not necessarily confirm that we have a project plugin.

File: SpatialGDK/Build/Scripts/BuildWorker.bat [hunk 2]
```diff
@@ -1,22 +1,69 @@
...
 
-rem Build Unreal project
-pushd "%~dp0..\..\..\..\..\"
+rem Try and build as a project plugin first, check for a project plugin structure.
+set UnrealProjectDir="%~dp0..\..\..\..\..\"
+set FoundUproject=""
 
+for /f "delims=" %%A in (' powershell -Command "(Get-ChildItem -Path "%UnrealProjectDir%" *.uproject).FullName" ') do set FoundUproject="%%A"
+
+if %FoundUproject%=="" (
```
Comment (m-samiec) [inline SpatialGDK/Build/Scripts/BuildWorker.bat]:
```suggestion
if %FoundUProject%=="" (
```

File: SpatialGDK/Build/Scripts/BuildWorker.bat [hunk 3]
```diff
@@ -1,22 +1,69 @@
 @echo off
 
-rem Generate worker configs
-pushd "%~dp0..\..\..\..\..\..\spatial"
-spatial worker build build-config
-popd
+rem Usage: <GameName> <Platform> <Configuration> <game.uproject> [-nocompile] <Additional UAT args>
```
Comment (m-samiec) [inline SpatialGDK/Build/Scripts/BuildWorker.bat]: What output does a user get if they execute this script with not params?

File: SpatialGDK/Build/Scripts/BuildWorker.bat [hunk 4]
```diff
@@ -1,22 +1,69 @@
...
 set BUILD_EXE_PATH="Plugins\UnrealGDK\SpatialGDK\Binaries\ThirdParty\Improbable\Programs\Build.exe"
+goto :Build
+
+
+:BuildAsEnginePlugin
+rem If we are running as an Engine plugin then we need a full path to the .uproject file!
+set UPROJECT=%4
+
+if not exist %UPROJECT% (
+	echo To use BuildWorker.bat with an Engine plugin installation you must provide a full path to your .uproject file.
```
Comment (m-samiec) [inline SpatialGDK/Build/Scripts/BuildWorker.bat]: ❤️

File: SpatialGDK/Build/Scripts/BuildWorker.bat [hunk 5]
```diff
@@ -1,22 +1,69 @@
...
+
+rem First build the spatial worker configs
+pushd "%SpatialDir%"
+spatial worker build build-config
+popd
+
+rem We pushd to the location of the project file to allow specifying just the .uproject name and not the full path (both are allowed for project plugins).
+pushd %UnrealProjectDir%
 
 if not exist %BUILD_EXE_PATH% (
```
Comment (m-samiec) [inline SpatialGDK/Build/Scripts/BuildWorker.bat]: Should this be moved to where it's set?

Comment (joshuahuburn) [reply SpatialGDK/Build/Scripts/BuildWorker.bat]: It's set in two locations and therefore we check here when it's going to be used.

File: SpatialGDK/Build/Scripts/BuildWorker.bat (hunk 3 above)
Comment (joshuahuburn) [reply SpatialGDK/Build/Scripts/BuildWorker.bat]: Added some parameter checking which prints if incorrect number of args are passed in

File: SpatialGDK/Build/Scripts/BuildWorker.bat [hunk 6]
```diff
@@ -1,22 +1,73 @@
...
-rem Generate worker configs
-pushd "%~dp0..\..\..\..\..\..\spatial"
-spatial worker build build-config
-popd
+if [%4] == [] goto :MissingParams
 
-rem Build Unreal project
-pushd "%~dp0..\..\..\..\..\"
+rem Try and build as a project plugin first, check for a project plugin structure.
+set UnrealProjectDir="%~dp0..\..\..\..\..\"
```
Comment (improbable-valentyn) [inline SpatialGDK/Build/Scripts/BuildWorker.bat]: Variable naming style is inconsistent (`%UnrealProjectDir%` vs `%BUILD_EXE_PATH%`)

File: SpatialGDK/Build/Scripts/BuildWorker.bat [hunk 7]
```diff
@@ -1,22 +1,73 @@
...
 set BUILD_EXE_PATH="Plugins\UnrealGDK\SpatialGDK\Binaries\ThirdParty\Improbable\Programs\Build.exe"
+goto :Build
+
+
+:BuildAsEnginePlugin
+rem If we are running as an Engine plugin then we need a full path to the .uproject file!
+set UPROJECT=%4
+
+if not exist %UPROJECT% (
+	echo To use BuildWorker.bat with an Engine plugin installation you must provide a full path to your .uproject file.
```
Comment (improbable-valentyn) [inline SpatialGDK/Build/Scripts/BuildWorker.bat]:
```suggestion
	echo To use BuildWorker.bat with an Engine plugin installation you must provide the full path to your .uproject file.
```
I may be wrong here though 😬

File: SpatialGDK/Build/Scripts/BuildWorker.bat [hunk 8]
```diff
@@ -1,22 +1,73 @@
...
+set SpatialDir="%~dp0..\..\..\..\..\..\spatial"
 set BUILD_EXE_PATH="Plugins\UnrealGDK\SpatialGDK\Binaries\ThirdParty\Improbable\Programs\Build.exe"
+goto :Build
+
+
+:BuildAsEnginePlugin
+rem If we are running as an Engine plugin then we need a full path to the .uproject file!
+set UPROJECT=%4
+
+if not exist %UPROJECT% (
```
Comment (m-samiec) [inline SpatialGDK/Build/Scripts/BuildWorker.bat]: This won't get his anymore right?

File: SpatialGDK/Build/Scripts/BuildWorker.bat [hunk 9]
```diff
@@ -1,22 +1,73 @@
...
+
+rem Path to the SpatialGDK build tool as an Engine plugin.
+set BUILD_EXE_PATH="%~dp0..\..\Binaries\ThirdParty\Improbable\Programs\Build.exe"
+set SpatialDir=%UnrealProjectDir%..\spatial\
+
+
+:Build
+
+if not exist %SpatialDir% (
+	echo Could not find the projects 'spatial' directory! Please ensure your input arguments are correct and your GDK is correctly installed.
```
Comment (improbable-valentyn) [inline SpatialGDK/Build/Scripts/BuildWorker.bat]:
```suggestion
	echo Could not find the project's 'spatial' directory! Please ensure your input arguments are correct and your GDK is correctly installed.
```

File: SpatialGDK/Build/Scripts/BuildWorker.bat (hunk 8 above)
Comment (improbable-valentyn) [reply SpatialGDK/Build/Scripts/BuildWorker.bat]: It would if you specify a non-existent uproject path.

Comment (joshuahuburn) [reply SpatialGDK/Build/Scripts/BuildWorker.bat]: Actually this is a necessary check, essentially if you are using an Engine Plugin you must be using a full path to the `.uproject`. If the `.uproject` doesn't exist (meaning a relative path has been passed). Then we print the error below :D

File: SpatialGDK/Build/Scripts/BuildWorker.bat (hunk 7 above)
Comment (joshuahuburn) [reply SpatialGDK/Build/Scripts/BuildWorker.bat]: Correct

**Pull request merged**